### DISCUSSION TITLE
resample from latlon large target grid to smaller grid

### DISCUSSION THREAD

**Discussion by marstonsward** [pytroll/pyresample#562] [Q&A] [answered]
I'm trying to resample and reproject some data from a global, regular latlon grid to much smaller LAEA grid, which is essentially going from a global gris defined in degrees to a smaller region defined in meters. The problem comes in here:

```python
result = kd_tree.resample_nearest(grid_def, da.values, area_def, radius_of_influence=50000, epsilon=0.5)
```
Code example
```python
import os
os.environ["PROJ_LIB"] = "/root/anaconda3/envs/exp/share/proj"
import json
import xarray as xr
import numpy as np
from pyresample.geometry import AreaDefinition, GridDefinition
from pyresample import kd_tree, save_quicklook

def prj4_string_to_dict(pj4str):
    ret = {}
    segs = pj4str.split(' ')
    for s in segs:
        if s == '': continue
        if '+units' in s:
            ret['units'] = s.split('=')[-1]
        elif '+proj' in s:
            ret['proj'] = s.split('=')[-1]
        elif '+lat_0' in s:
            ret['lat_0'] = s.split('=')[-1]
        elif '+lon_0' in s:
            ret['lon_0'] = s.split('=')[-1]
        elif '+a' in s:
            ret['R'] = s.split('=')[-1]
        elif '+ellps' in s:
            ret['ellps'] = s.split('=')[-1]
        elif '+over' in s:
            ret['over'] = s.split('=')[-1]
        else:
            print(f'No logic for {s}')
            print(segs)
    return ret
# Load the data array from the netCDF file
da = xr.open_dataarray('gfs_orography.nc', engine='netcdf4')
lon2d, lat2d = np.meshgrid(da.coords['lon'].values, da.coords['lat'].values)
# there it is required that the coordinates arrays are the same shape, which necessitates the following: 
grid_def = GridDefinition(lons=lon2d, lats=lat2d)

# Load the region projection info from a JSON file
with open('rois.json', 'r') as json_file:
    region_config = json.load(json_file)

for roi, info in region_config.items():
    x_len = info["xLen"]
    y_len = info["yLen"]
    resolution = info['defaultBin']
    projString = info['projection']

    # Calculate coordinates for the sector polygon
    x_min, x_max = -x_len / 2, x_len / 2
    y_min, y_max = -y_len / 2, y_len / 2

    area_extent = (x_min, y_min, x_max, y_max)
    area_id = regionName
    description = regionName
    proj_id = regionName
    projection = prj4_string_to_dict(projString)
    width = x_len
    height = y_len
    area_def = AreaDefinition(area_id, description, proj_id, projection, width, height, area_extent)
    print(area_def)
    result = kd_tree.resample_nearest(grid_def, da.values, area_def, radius_of_influence=50000, epsilon=0.5)
```
Error:

```bash
Exception has occurred: _ArrayMemoryError       (note: full exception trace is shown but execution is paused at: <module>)
Unable to allocate 33.8 TiB for an array with shape (37158912000000,) and data type bool
  File "/home/synrad/sw/mlde-rts/shapefile_extract_slice.py", line 74, in <module> (Current frame)
    result = kd_tree.resample_nearest(grid_def,
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
numpy.core._exceptions._ArrayMemoryError: Unable to allocate 33.8 TiB for an array with shape (37158912000000,) and data type bool
``` 

Why is it pulling so much memory? Can I define the source grid by just using 1D coords? Is there a way to do that? The shape of the source data is: 741, 1400 and the target region is: extent: (-4032000.0, -2304000.0, 4032000.0, 2304000.0) in meters.

**Comment by djhoese** [comment]
Side suggestion: You can get rid of the projection string -> dict conversion. AreaDefinitions understand anything pyproj CRS objects understand so WKT, EPSG, PROJ.4 strings, PROJ.4 dictionaries, or CRS objects themselves should all work.

For your GridDefinition, you don't need to define it as a GridDefinition. If these are equally spaced pixels (which they are) then creating an AreaDefinition with the equivalent CRS (projection) and calculating extents in degrees should be fine. Note the extents are the outer edges of the pixels, not the centers of the pixels. This should at least save you the memory for the creation of the GridDefinition, but...what is the width and height of your target Area? Are you sure on the size of your input data? Can you print out all the shapes involved in this code?

**Reply by marstonsward** [reply]
@djhoese Thanks for the quick response and tips. I removed the function and applied your suggestions but got the same error.

![image](https://github.com/pytroll/pyresample/assets/1478822/862e2a5f-1241-4ffe-8e59-926d8ed32016)

updated code:
```python
import os
os.environ["PROJ_LIB"] = "/root/anaconda3/envs/exp/share/proj"
import matplotlib.pyplot as plt
import json
import xarray as xr
import numpy as np
from pyresample.geometry import AreaDefinition, GridDefinition
from pyresample import kd_tree, save_quicklook

# Load the data array from the netCDF file
da = xr.open_dataarray('/root/gfs_orography.nc', engine='netcdf4')
shape = da.shape
area_id = 'globe'
proj_string = '+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs'
area_extent = (-180.0, -90.0, 180.0, 90.0)
shape = shape
src_def = AreaDefinition.from_extent(area_id, proj_string, shape, area_extent)

# Load the region projection info from a JSON file
with open('gsw_rts_map_coords/domains.json', 'r') as json_file:
    region_config = json.load(json_file)

for regionName, info in region_config.items():
    # Extract parameters from the definition
    x_len = info["xLen"]
    y_len = info["yLen"]
    resolution = info['defaultBin']
    projString = info['projection']
    # Calculate coordinates for the sector polygon
    x_min, x_max = -x_len / 2, x_len / 2
    y_min, y_max = -y_len / 2, y_len / 2

    area_extent = (x_min, y_min, x_max, y_max)
    area_id = regionName
    description = f'GSWR {regionName}'
    proj_id = regionName
    projection = projString
    width = x_len
    height = y_len
    targ_def = AreaDefinition(area_id, description, proj_id, projection, width, height, area_extent)
    
    result = kd_tree.resample_nearest(src_def, da.values, targ_def, radius_of_influence=50000, epsilon=0.5)
```

But I think the problem is the source data size: 
```
Area ID: my_area
Description: my_area
Projection ID: my_area
Projection: {'a': '6370997.0', 'ellps': 'sphere', 'lat_0': '50', 'lon_0': '-135', 'over': 'True', 'proj': 'laea', 'type': 'crs', 'units': 'm'}
Number of columns: 8064000
Number of rows: 4608000
Area extent: (-4032000.0, -2304000.0, 4032000.0, 2304000.0)
```
I see that PyResample is trying to make a grid of the target grid. Can this be avoided?

**Reply by djhoese** [reply]
What is the `my_area` you've printed out in the above comment? You said your source data is 721 rows by 1440 columns, but the area you've printed out with the comment "I think the problem is the source data size" has 4608000 rows by 8064000 columns. Your source data and source area should be the same size.

Additionally, that number of rows and columns is **HUGE**:

```
In [1]: 8064000 * 4608000
Out[1]: 37158912000000
```

That's 37 trillion elements. For 32-bit floats (x and y coordinates usually need 64-bit) that's something like 135TB's per coordinate array...right? That target area definition has a resolution of 1 meter per pixel:

```
In [5]: (4032000.0 - -4032000.0) / 8064000
Out[5]: 1.0
```

Could you explain again what you're trying to do?

**Reply by djhoese** [reply, marked as answer]
And no, pyresample needs to compute all the coordinates for these areas so it can:

1. Build a KDTree of all input pixel coordinates
2. Query the KDTree to say "what pixel in the input data is nearest to this output pixel"

**Reply by marstonsward** [reply]
Oh man! I forgot to divide the weight and width by the resolution. Now it works like a charm 👍 I was so focused on understanding PyResample that I forgot that part.
Thanks for the extra eyes and tips.

**Reply by marstonsward** [reply]
@djhoese 

One last question: can `save_quicklook()` add the labels to meridians and parallels in the plot?

**Reply by djhoese** [reply]
Glad it was a simple fix. I was really concerned you actually needed that size/resolution.

I honestly didn't know `save_quicklook` existed. It looks like it is just using cartopy so it should be possible, but I don't see any option for it. You may be able to take the code here and modify it as needed (or make a pull request):

https://github.com/pytroll/pyresample/blob/78b19ca19aea2c030d6ff3d9ebb296301528c6d9/pyresample/plot.py#L241-L269

**Reply by marstonsward** [reply]
@djhoese 

I think I have solved the grid lines labels but I have to do the PR next week. I have another issue.
It looks like PyResample has issues when the target grid crosses the dateline. This, if memory serves me, is a know issue. Is there a solution  to handle projecting across the dateline?

**Reply by marstonsward** [reply]
Here's an example of the region: 

`+units=m +proj=laea +lat_0=-50 +lon_0=-135 +a=6370997.0 +ellps=sphere +over=True`
Now you just need to choose a size that makes this crosses the dateline. Ex: x = 2500 and y = 2250.

**Reply by djhoese** [reply]
The `+over=True` is technically deprecated and not preferred in modern versions of PROJ. Strictly speaking if you have an AreaDefinition that crosses the anti-meridian of that projection then it is not really a valid contiguous space on that coordinate system/projection. There are ways to get around this, but it also depends on your exact use case. Could you provide the current version of your code and area definition? Technically speaking I think the lon/lat anti-meridian should be fine, but the crossing of your own area's anti-meridian is another story.

One option is to do `+pm=180` (or whatever number you want), which basically means the geographic prime meridian of my projection is actually at longitude 180. This is kind of a weird hack in my opinion but it does work when it is needed. It may not actually be needed. I just need more information about the target area you're producing now and your input data.

**Reply by marstonsward** [reply]
@djhoese 

Aha...this is some serious legacy code written many years ago and has not been updated with time. I'm really a newbie in the CRS projection world so I have many blind spots to fill in. I appreciate you talking the time to look at this.

My goal is to `pythonize` this code and bring it into the modern era. Your tip to remove `+over` helped greatly but I

```bash
targ_def
Area ID: sector00T
Description: GSWR sector00T
Projection ID: sector00T
Projection: {'ellps': 'sphere', 'lat_0': '0', 'lon_0': '-135', 'no_defs': 'None', 'proj': 'laea', 'type': 'crs', 'units': 'm', 'x_0': '0', 'y_0': '0'}
Number of columns: 2544
Number of rows: 2016
Area extent: (-5088000.0, -4032000.0, 5088000.0, 4032000.0)

 mamba list pyresample
# packages in environment at /root/anaconda3/envs/xesmf_env:
#
# Name                    Version                   Build  Channel
pyresample                1.27.1          py311h320fe9a_2    conda-forge
pyproj                        3.6.1            py311h1facc83_4    conda-forge
```

Code:

```python
import os
os.environ["PROJ_LIB"] = "/root/anaconda3/envs/exp/share/proj"

import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import json
import numpy as np
from pyproj import crs
import geoxarray
import xarray as xr
from pyresample.geometry import AreaDefinition
from pyresample import kd_tree, bilinear

# Load the data array from the netCDF file
da_src = xr.open_dataarray('/root/gfs_orography.nc', engine='netcdf4')
shape = da_src.shape
area_id = 'globe'
proj_string = '+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs'
area_extent = (-180.0, -90.0, 180.0, 90.0)
shape = shape
src_def = AreaDefinition.from_extent(area_id, proj_string, shape, area_extent)

# Load the region projection info from a JSON file
with open('domains.json', 'r') as json_file:
    region_config = json.load(json_file)

for regionName, info in region_config.items():
    if 'sector' not in regionName: continue
    print(f'Extract parameters from the definition: {regionName}')
    x_len = info["xLen"]
    y_len = info["yLen"]
    resolution = info['defaultBin']
    projString = info['projection']
    if 'ellps' not in projString:
        projString = f'{projString}+ellps=sphere +type=crs'
    # Calculate coordinates for the sector polygon
    x_min, x_max = -x_len / 2, x_len / 2
    y_min, y_max = -y_len / 2, y_len / 2

    area_extent = (x_min, y_min, x_max, y_max)
    description = f'GSWR {regionName}'
    proj_id = regionName
    projection = projString
    # Get the array shape
    width = int(x_len // resolution)
    height = int(y_len // resolution)
    # set target grid
    targ_def = AreaDefinition(regionName,
                              description,
                              proj_id,
                              projection,
                              width,
                              height,
                              area_extent)

    result = kd_tree.resample_nearest(src_def,
                                      da_src.values,
                                      targ_def,
                                      radius_of_influence=25000,
                                      epsilon=0.5)

    # create the dataarray in projected CRS
    # Create the coordinates
    x_coords = np.linspace(x_min, x_max, width)
    y_coords = np.linspace(y_min, y_max, height)

    da = xr.DataArray(result, dims=('y', 'x'), coords={'y': y_coords, 'x': x_coords})
    da.name = regionName

    # Plotting
    fig = plt.figure(figsize=(10, 6))
    ax = plt.axes(projection=targ_def.to_cartopy_crs())
    ax.coastlines(resolution='110m')
    ax.set_global()
    print('Plot DataArray')
    # Not sure why this is necessary but it is only way I could get the data N-S
    da.values = np.flipud(da.values)

    cbar_kwargs = {'orientation':'vertical', 'shrink': 0.78, 'pad':0.05}
    col = da.plot(ax=ax,transform=targ_def.to_cartopy_crs(),
                  robust=True, cbar_kwargs=cbar_kwargs, cmap='RdBu_r')
    gl = ax.gridlines(draw_labels=True,
                      linewidth=1,
                      color='gray',
                      alpha=0.5,
                      linestyle='--')
    gl.xlabels_top = False
    gl.ylabels_right = False

    plt.savefig(f'{regionName}.png',
                dpi=300, bbox_inches='tight')
    plt.close()
```

I'm going to move from proj4 to WKT2 or maybe find a better way to define and store the regions I'm working with. I think the problem was the `+over`.

**Reply by djhoese** [reply]
Which region (or what's one) that you're having trouble with? What is the area information generated for it? After removing `+over` are you still having trouble?

**Reply by marstonsward** [reply]
It looks like it works now.
I'm trying to use Xarray Bilinear resampler now. But the docs seem incomplete. 
Question: can pyresample use a wkt instead of a proj4 string?

**Reply by marstonsward** [reply]
The region is given in the first part of the code: sector00T.

**Reply by djhoese** [reply]
> I'm trying to use Xarray Bilinear resampler now. But the docs seem incomplete.

I would highly recommend not using this resampler and instead try using the nearest neighbor or gradient_search. I am currently trying to find time to migrate all satpy-xarray-friendly resampler classes to pyresample. Otherwise, the xarray bilinear resampler has very poor performance. It works, but not well. And yes, these classes are overall poorly documented as they were added as-needed by Satpy developers as a "proof of concept" but saw very little proper maintenance and documentation. This is what I'm working on now.

> Question: can pyresample use a wkt instead of a proj4 string?

Yes. Anything that pyproj's CRS object can handle (which includes WKT) should work when passed to an AreaDefinition.

> The region is given in the first part of the code: sector00T.

Ah. Duh. Ok let me take a look at that area.

**Reply by djhoese** [reply]
Yeah, I don't see anything odd. Let's just hope that `+over` was being too smart for its own good with modern PROJ/pyproj.

**Comment by marstonsward** [comment]
@djhoese 

I've gotten the NN to work very well, but ran into a problem with gradient_search for bilinear interpolation. In my case I'm going from `Areadefinition` to `Areadefinition` with an input data that is 2D (same as above and which works for the NN example). Here's my function: 

```python
from pyresample.geometry import AreaDefinition
from pyresample import kd_tree, gradient
from pyresample import load_area

def gradient_bilinear(source_def, target_def, data):
    resampler = gradient.create_gradient_search_resampler(source_def, target_def)
    regrid = resampler(data)
    return regrid

print(f'Extract parameters from the definition: {regionName}')
targ_def = load_area('/home/rois.yaml', regionName)

# convert to a double
data = da_src.astype('float64')
result = gradient_bilinear(src_def, targ_def,  data.values)
```
But got the following error:

```bash
Exception has occurred: TypeError
'ResampleBlocksGradientSearchResampler' object is not callable
  File "sector_reprojection.py", line 32, in gradient_bilinear
    regrid = resampler(data)
             ^^^^^^^^^^^^^^^
 ```

Am I doing this correct? Not much help in the docs on this method.

**Reply by djhoese** [reply]
I believe the gradient search is still relatively experimental. I'm also not seeing your full traceback in your comment so can't guess as to what is going on. Lastly, what is the object type of `da_src`?

CC @mraspaud the gradient search developer

**Reply by marstonsward** [reply]
@djhoese This is the only traceback I got from VS Code. I was running in debug mode. I looked into the pyresample code (it's a bit more professional than I'm used to - lot of decorators and what not) but I believe that the call is simply (from the gradient code dir):

```python
def create_gradient_search_resampler(source_geo_def, target_geo_def):
    """Create a gradient search resampler."""
    if isinstance(source_geo_def, AreaDefinition) and isinstance(target_geo_def, AreaDefinition):
        return ResampleBlocksGradientSearchResampler(source_geo_def, target_geo_def)
    elif isinstance(source_geo_def, SwathDefinition) and isinstance(target_geo_def, AreaDefinition):
        return StackingGradientSearchResampler(source_geo_def, target_geo_def)
    raise NotImplementedError
```
This should return a class instance `ResampleBlocksGradientSearchResampler(source_geo_def, target_geo_def)` but it seems this is broken.

`da_src` is a simple `dataarray` of the source model data that I'm trying to reproject from `latlong` to `LAEA`. 
Apologies for the incomplete code. My company doesn't want proprietary info posted so I have to be careful.

**Reply by djhoese** [reply]
Just as an experiment/test, can you try creating the resampler yourself by calling `ResampleBlocksGradientSearchResampler` yourself with your two areas (source and target areas). I'm not sure how it is possible that this Resampler class "is not callable" as the error says.

**Reply by marstonsward** [reply]
Hi,

I got the same error even if I import the module directly from gradient and bypassed the create wrapper as you suggested:

```python
from pyresample.geometry import AreaDefinition
from pyresample import kd_tree, gradient
from pyresample import load_area
import xarray as xr
from pyresample.gradient import ResampleBlocksGradientSearchResampler

def gradient_bilinear(source_def, target_def, data):
    #resampler = gradient.create_gradient_search_resampler(source_def, target_def)
    resampler = ResampleBlocksGradientSearchResampler(source_def, target_def)
    regrid = resampler(data)
    return regrid

# Load the data array from the netCDF file
da_src = xr.open_dataarray('/root/gfs_orography.nc', engine='netcdf4')
shape = da_src.shape
area_id = 'globe'
proj_string = '+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs'
area_extent = (-180.0, -90.0, 180.0, 90.0)
shape = shape
src_def = AreaDefinition.from_extent(area_id, proj_string, shape, area_extent)

targ_def = load_area('/root/rois.yaml', 'sector1')

# Cython need this converted to a double
data = da_src.astype('float64')
result = gradient_bilinear(src_def, targ_def, data.values)
```

This time I've even run this program outside of VS Code to ensure this is not a VS Code things. I got very little traceback and the same error:

```bash
Traceback (most recent call last):
  File "/root/sector_reprojection.py", line 81, in <module>
    result = gradient_bilinear(src_def, targ_def, data.values)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/sector_reprojection.py", line 34, in gradient_bilinear
    regrid = resampler(data)
             ^^^^^^^^^^^^^^^
TypeError: 'ResampleBlocksGradientSearchResampler' object is not callable
```

**Reply by marstonsward** [reply]
I maybe wrong as I'm certainly not up on the latest python coding styles, but in the `__init__.py` file in the `gradient` directory, shouldn't 
`ResampleBlocksGradientSearchResampler` and `StackingGradientSearchResampler` be imported from some module?
 that is:

```python
from pyresample.resampler import BaseResampler,  resample_blocks

should be something like this:

from pyresample.resampler import (
    BaseResampler,
    resample_blocks,
    ResampleBlocksGradientSearchResampler,
    StackingGradientSearchResampler
)
```

**Reply by djhoese** [reply]
All of that code currently lives here: https://github.com/pytroll/pyresample/blob/main/pyresample/gradient/__init__.py so that is not the problem.

With your VSCode error it is much clearer that the problem isn't the creation of the resampler but the usage of the resampler. You're calling `resampler(data)`, but I don't think there is a `__call__` method on the resampler. You need to call `resampler.resample(data)`. Try that and let's see how it goes.

**Reply by marstonsward** [reply]
Hmm...I had not thought that the resampler was the problem.

However, after testing the new call:

```python
def gradient_bilinear(source_def, target_def, data):
    da = data.rename({'lat': 'y', 'lon': 'x'})
    resampler = gradient.create_gradient_search_resampler(source_def, target_def)
    regrid = resampler.resample(da)
    return regrid
```

```bash
Exception has occurred: AttributeError
'numpy.ndarray' object has no attribute 'rechunk'
  File "sector_reprojection.py", line 33, in gradient_bilinear
    regrid = resampler.resample(da)
             ^^^^^^^^^^^^^^^^^^^^^^
  File "sector_reprojection.py", line 80, in <module>
    result = gradient_bilinear(src_def, targ_def, data)
AttributeError: 'numpy.ndarray' object has no attribute 'rechunk'```

**Reply by marstonsward** [reply]
I did manage to get his to work with this following:

```python
def gradient_bilinear(source_def, target_def, data):
    da = data.rename({'lat': 'y', 'lon': 'x'}).chunk(chunks='auto')
    resampler = gradient.create_gradient_search_resampler(source_def, target_def)
    regrid = resampler.resample(da)
    return regrid
```

But this definition doesn't work, i.e., no data was reprojected
```yaml
sector04T:
  area_extent:
    lower_left_xy:
    - -5088000
    - -4032000
    upper_right_xy:
    - 5088000
    - 4032000
  area_id: 's04'
  description: 'region'
  projection:
    a: 6370997
    ellps: 'sphere'
    lat_0: 0
    lon_0: 140
    proj: 'laea'
    units: 'm'
  shape:
    height: 2016
    width: 2544
```
![image](https://github.com/pytroll/pyresample/assets/1478822/b674dfbe-4eef-4b2a-9139-5d77664b4cda)

But this area definition works:

```yaml
sector03T:
  area_extent:
    lower_left_xy:
    - -5088000
    - -4032000
    upper_right_xy:
    - 5088000
    - 4032000
  area_id: 's03'
  description: 'region'
  projection:
    a: 6370997
    ellps: 'sphere'
    lat_0: 0
    lon_0: 65
    proj: 'laea'
    units: 'm'
  shape:
    height: 2016
    width: 2544
```
![image](https://github.com/pytroll/pyresample/assets/1478822/ac11c881-eaf2-4703-bb29-b71f073679bd)


I suspect this has to do with the antimeridian. Is this a bug? I', using the PyResample area definition format from a yaml file. It seems this bilinear method needs something like `+over=True`.


Also getting failures on full disk GOES projections such as:

```yaml
GE_geos:
  area_extent:
    lower_left_xy:
    - -5568000
    - -5568000
    upper_right_xy:
    - 5568000
    - 5568000
  area_id: 'GE_geos'
  description: 'region'
  projection:
    ellps: 'WGS84'
    h: 35785831
    lon_0: -75
    proj: 'geos'
    units: 'm'
  shape:
    height: 3712
    width: 3712
```

**Reply by djhoese** [reply]
I haven't tested gradient search much. @pnuu @mraspaud do you know if there are any known issues with projections that overlap the lon/lat anti-meridian?

**Reply by marstonsward** [reply]
@djhoese 

Is there no traction on this issue from the developers?

**Reply by djhoese** [reply]
You've caught the other maintainers at a bad time. They all went off on their holiday breaks and aren't back yet. I've linked them to my comment above in the pytroll slack so they should see it when they get back.

**Reply by marstonsward** [reply]
Yeah, I used to work at SMHI many years ago. I know how generous their vacation plan is :-) 
Understood. Looking forward to getting this solved.

**Reply by mraspaud** [reply]
I'll have a look

**Reply by mraspaud** [reply]
@Marston Looks like the resample_blocks-based resampler need to be called a bit differently. I successfully resampled data over both sector03T and sector04T using a modified `gradient_bilinear` as follows:
```python
def gradient_bilinear(source_def, target_def, data):
    resampler = gradient.create_gradient_search_resampler(source_def, target_def)
    resampler.precompute()
    regrid = resampler.compute(data, method="bilinear")
    return regrid
```

**Reply by marstonsward** [reply]
@mraspaud 

Appreciate the update. Unfortunately, the suggested changes are still not working for me. Here's the modified function:

```python
def gradient_bilinear(source_def, target_def, data):
    da = data.rename({'lat': 'y', 'lon': 'x'}).chunk(chunks='auto')
    resampler = gradient.create_gradient_search_resampler(source_def, target_def)
    resampler.precompute()
    return resampler.compute(da, method='bilinear')
```

Here are the definitions of a few sectors where it fails completely (without an error message):

```yaml
sector04T:
  area_extent:
    lower_left_xy:
    - -5088000
    - -4032000
    upper_right_xy:
    - 5088000
    - 4032000
  area_id: 'sector04T'
  description: 'PyResample definition of a region'
  projection:
    a: 6370997
    ellps: 'sphere'
    lat_0: 0
    lon_0: 140
    proj: 'laea'
    units: 'm'
  shape:
    height: 2016
    width: 2544
GE_geos:
  area_extent:
    lower_left_xy:
    - -5568000
    - -5568000
    upper_right_xy:
    - 5568000
    - 5568000
  area_id: 'GE_geos'
  description: 'PyResample definition of a region'
  projection:
    ellps: 'WGS84'
    h: 35785831
    lon_0: -75
    proj: 'geos'
    units: 'm'
  shape:
    height: 3712
    width: 3712
GW_geos:
  area_extent:
    lower_left_xy:
    - -5568000
    - -5568000
    upper_right_xy:
    - 5568000
    - 5568000
  area_id: 'GW_geos'
  description: 'PyResample definition of a region'
  projection:
    ellps: 'WGS84'
    h: 35785831
    lon_0: -137
    proj: 'geos'
    units: 'm'
  shape:
    height: 3712
    width: 3712
H8_geos:
  area_extent:
    lower_left_xy:
    - -5568000
    - -5568000
    upper_right_xy:
    - 5568000
    - 5568000
  area_id: 'H8_geos'
  description: 'PyResample definition of a region'
  projection:
    ellps: 'WGS84'
    h: 35785831
    lon_0: 140.7
    proj: 'geos'
    units: 'm'
  shape:
    height: 3712
    width: 3712
```

**Reply by mraspaud** [reply]
ok, to make sure this is not a problem with the data, could you share a data file that I work with?

**Reply by marstonsward** [reply]
No problem. Here it is: 
[gfs_orography.tar.gz](https://github.com/pytroll/pyresample/files/13888575/gfs_orography.tar.gz)

**Reply by mraspaud** [reply]
Ok, so I can reproduce your problem with your data. Putting `over: true` in the yaml definition of sector04T makes some of the data visible, but it's not fully covered... I need to investigate more.

**Reply by marstonsward** [reply]
If I add `over=True`, it fails for the KD Tree NN method. But as you have seen, it only solves part of the problem.

**Reply by mraspaud** [reply]
I think I found the problem for gradient resampling: the resampler uses shapely to find the overlap between the source area and destination area. However, shapely doesn't work well with the data line in particular, so the fact the source area is defined between -180 and 180 degrees is a problem. I guess these functions need reworking: https://github.com/pytroll/pyresample/blob/main/pyresample/slicer.py#L152-L188 . I think @ghiggi is working on a fix for this though.

A workaround for this would be to "re-center" the data somehow around the dateline and use not degrees but meters for the area extend units.

**Reply by marstonsward** [reply]
@mraspaud Thanks for looking into this. I understand about the dateline, and I can easily keep the source data in a 0:360 longitude grid. But I'm not sure what you mean by the units. How do this factor into the problem?

**Reply by mraspaud** [reply]
I'm not sure pyresample will handle 0:360 gracefully, as it wants to have -180:180. Hence using plate-carree with meters as a unit should work around the problem, I thin.

**Reply by marstonsward** [reply]
Let me see if I am understanding correctly: it the input data in on a plate-carree grid and expressed in meters not degrees, then this resampling with reprojection (LAEA) should work?
